### PR TITLE
Fixes #15046 - Add bmc_nic_setup snippet

### PR DESCRIPTION
**Pull request by bagasse**
Template part of http://projects.theforeman.org/issues/15046.

### CODE REVIEW

**Comment by mmoll** [comment]
This is looking very useful, but I don't have any hardware to test this, so only two comments from my side:
- Please make this snippet more OS agnostic. Have a look onto the puppet_setup snippet as an example. I'd like to see that working also on Debian/Ubuntu, idealy also on FreeBSD. Especially regarding the later (no GNU grep or awk) I'd recommend e.g. to use cut instead of awk and `tail -n 1` instead of `tail -1`
- I _think_ there's also hardware on the market that doesn't use 1 as its default channel (@lzap I think you came across some IBM h/w where that was the case), so it might make sense to extract that into a parameter.

**Comment by bagasse** [comment]
@mmoll, thank you for your comment. I've only dell hardware so i didn't test it on other hardware. I've modified the snippet to:
- Manage channel IPMI channel ID and use 1 by default
- Use package manager from os type family (maybe i have to manage dnf for recent fedora releases ?).
- Use portable options of grep, and replaced awk with cut

Modifications tested on CentOS7 at this time. No hardware test server free at this time, it will be difficult to me to test it on some debian or freebsd versions.

**Comment by lzap** [comment]
I also don't have IPMI enabled hardware myself, but since this is a snippet, I think we can merge it.

File: snippets/bmc_nic_setup.erb [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<%#
+kind: snippet
+name: bmc_nic_setup
+oses:
```
Comment (mmoll) [inline snippets/bmc_nic_setup.erb]: snippets don't have OS associations, so this should get deleted

File: snippets/bmc_nic_setup.erb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+- RedHat 7
+%>
+<%#
+# bmc_nic_setup snippet
+# This snippet configure BMC interfaces.
+# At this time, only IPMI provider is supported.
+# It was tested on some Dell hardware with CentOS 6 - CentOS 7 and should be compatible with respective
+# upstream RHEL versions.
+#
+# What this snippet do:
```
Comment (mmoll) [inline snippets/bmc_nic_setup.erb]: does

File: snippets/bmc_nic_setup.erb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+# * Set the authentication mecanism to 'MD5' for the used privilege level
+#
+# * Create or Update the provided user with provided privileges level. ADMINISTRATOR privilege level is used by default
+#   because ipmitool utility request ADMINISTRATOR privileges level if no privilege level is provided.
+#   If you want to use a lower level of privilges (like OPERATOR), don't forget to add the privilege level to request
+#   with your ipmi command (eg: ipmitool -I lanplus -L operator -U user1 -H 192.168.1.2 chassis status).
+#
+# * Enable the user created/updated
+#
+# What this snippet don't do:
```
Comment (mmoll) [inline snippets/bmc_nic_setup.erb]: doesn't

File: snippets/bmc_nic_setup.erb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+os_name   = @host.operatingsystem.name
+
+%>
+
+<% if bmc_nic && bmc_nic.managed? && bmc_nic.provider == "IPMI" && !bmc_nic.username.empty?  && !bmc_nic.password.empty? -%>
+
+# Configure BMC interface
+<% case os_family %>
+<% when 'Redhat' then %>
+yum  -t -y install ipmitool
```
Comment (mmoll) [inline snippets/bmc_nic_setup.erb]: one space too much here

File: snippets/bmc_nic_setup.erb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+privilege_levels = {'ADMINISTRATOR' => 4, 'OPERATOR' => 3}
+user_privilege_level = @host.params['bmc_ipmi_priv_level'] || 'ADMINISTRATOR'
+auth_method = @host.params['bmc_ipmi_auth_method'] || 'MD5'
+channel_id = @host.params['bmc_ipmi_channel_id'] || '1'
+
+os_family = @host.operatingsystem.family
+os_major  = @host.operatingsystem.major.to_i
+os_name   = @host.operatingsystem.name
+
```
Comment (mmoll) [inline snippets/bmc_nic_setup.erb]: major and name are not used, AFAICS?

**Comment by mmoll** [comment]
There's no dnf usage in any of the templates, yet, so that's OK

**Comment by bagasse** [comment]
Done.

**Comment by mmoll** [comment]
merged, merci @bagasse!

**Comment by bagasse** [comment]
@mmoll, @lzap, thank you for comments/reviews

**Pull request merged**